### PR TITLE
#36 adopt keycloak tokens

### PR DESCRIPTION
**Pull request by rgraf**
Adopted JWT token user authorization.

### CODE REVIEW

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
+		throws ApplicationAuthenticationException;
+
+    /**
+     * This method adopts KeyCloack token from HTTP request
+     * @param request The HTTP request
+     * @return list of Authentication objects
+     * @throws ApplicationAuthenticationException
+     * @throws ApiKeyExtractionException
+     */
+    public List<? extends Authentication> processJwtToken(HttpServletRequest request)
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java]: we don#t need this method in the interface

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+	    throws ApplicationAuthenticationException, ApiKeyExtractionException, AuthorizationExtractionException;
+    
+    /**
+     * This method verifies write access rights for particular api and operation
+     * @param authenticationList The list of authentications extracted from the JWT token
+     * @param operation The name of current operation
+     * @return true if authenticated, false otherwise
+     * @throws ApplicationAuthenticationException
+     */
+    public boolean authorizeWriteAccess(List<? extends Authentication> authenticationList, String operation) throws ApplicationAuthenticationException;
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java]: this method should take the request as input ...

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java [hunk 3]
```diff
@@ -1,69 +1,135 @@
...
+	
+	//check if null
+	if (wsKey == null)
+	    throw new ApplicationAuthenticationException(I18nConstants.MISSING_APIKEY, I18nConstants.MISSING_APIKEY,
+		    null, HttpStatus.UNAUTHORIZED, null);
+	//check if empty
+	if (StringUtils.isEmpty(wsKey))
+	    throw new ApplicationAuthenticationException(I18nConstants.EMPTY_APIKEY, I18nConstants.EMPTY_APIKEY, null);
+	//validate api key
+	if (!wsKey.equals("apidemo")) {
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: This check was removed in the develop branch, please merge first the latest develop version into this branch..

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java [hunk 4]
```diff
@@ -1,69 +1,135 @@
...
+			new String[] { wsKey });
+	    }
+	}
+    }
+
+     
+    /* (non-Javadoc)
+     * @see eu.europeana.api.commons.service.authorization.AuthorizationService#authorizeWriteAccess(java.util.List, java.lang.String, java.lang.String)
+     */
+    public boolean authorizeWriteAccess(List<? extends Authentication> authenticationList, String operation)
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: need to refactor and take the request object as input

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java [hunk 5]
```diff
@@ -1,69 +1,135 @@
...
+    
+    /**
+     * This method verifies write access rights for particular api and operation
+     * @param authenticationList The list of authentications extracted from the JWT token
+     * @param api The current api name
+     * @param operation The name of current operation
+     * @return true if authenticated, false otherwise
+     * @throws ApplicationAuthenticationException
+     */
+    public static boolean checkPermissions(List<? extends Authentication> authenticationList, String api, String operation)
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: doesn't nneed to be public, protected is enough

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java [hunk 6]
```diff
@@ -1,69 +1,135 @@
...
+    public static boolean checkPermissions(List<? extends Authentication> authenticationList, String api, String operation) 
+	    throws ApplicationAuthenticationException {
+	 boolean res = false;
+	 
+	 for(Authentication authentication : authenticationList) {
+	     String authenticationApi = (String) authentication.getDetails();
+	     @SuppressWarnings("unchecked")
+	     List<GrantedAuthority> authorityList = (List<GrantedAuthority>) authentication.getAuthorities();
+	     if(authenticationApi.equals(api)) {
+		 for(GrantedAuthority authority : authorityList) {
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: this for loop should be extracted to own method throug refactoring

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java [hunk 7]
```diff
@@ -1,69 +1,135 @@
...
+//	     } else {
+//	         throw new ApplicationAuthenticationException(I18nConstants.INVALID_API_NAME, I18nConstants.INVALID_API_NAME, null);
+	     }
+	     
+	     if(res) {
+		 break;
+	     }
+	 }
+	 
+	 return res;
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: the method should return void and throw and exception if the user doesn't have the permissions to perform the required operation

File: commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/Operations.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+	//search is also retrieve
+	//crud operations
+	public static final String RETRIEVE = "retrieve";
+	public static final String CREATE = "create";
+	public static final String UPDATE = "update";
+	public static final String DELETE = "delete";
+	
+	//admin
+	public static final String ADMIN_ALL = "admin_all";
+	public static final String ADMIN_UNLOCK = "admin_unlock";
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/Operations.java]: only the CRUD and ADMIN_ALL are generic permission to stay within the api-commons modules. The rest of them are api specific and must not be included here..

File: commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package eu.europeana.api.commons.web.model.vocabulary;
+
+public enum UserRoles {
+
+	USER(new String[]{Operations.RETRIEVE, Operations.CREATE, Operations.DELETE, Operations.UPDATE}), 
+	ADMIN(new String[]{Operations.RETRIEVE, Operations.CREATE, Operations.DELETE, Operations.UPDATE, Operations.ADMIN_ALL, Operations.ADMIN_UNLOCK, Operations.ADMIN_REINDEX, Operations.USER_CREATE, Operations.USER_DELETE});
```
Comment (gsergiu) [inline commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java]: The user roles are API specific, they must not be included in the API commons, eventually an interface with the methods getName() and getPermissions() can be included in API commons

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+     * 
+     */
+    private static final long serialVersionUID = -3956175951431237501L;
+
+    Collection<? extends GrantedAuthority> collection = null;
+    
+    String details = null;
+    
+    
+    public EuropeanaAuthenticatonToken(Collection<? extends GrantedAuthority> arg0) {
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java]: Added constructor, which takes also details as an input

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	super(arg0);
+	// TODO Auto-generated constructor stub
+    }
+
+    @Override
+    public Object getCredentials() {
+	// TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java]: Generated comments removed

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	return null;
+    }
+
+    @Override
+    public Object getPrincipal() {
+	// TODO Auto-generated method stub
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java]: Method getPrincipal returns user name from the JWT token

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java [hunk 13]
```diff
@@ -54,7 +64,58 @@ public static String extractApiKey(HttpServletRequest request) throws ApiKeyExtr
...
+		throw new AuthorizationExtractionException("Custom data not available in provided JWT token");
+	    }
+
+	    // for each API in resource_access should be produced EuropeanaAuthenticationToken
+	    EuropeanaAuthenticatonToken authenticationToken;
+	    Collection<GrantedAuthority> rolesCollection; 
+	    for (Map.Entry<String, Object> entry : resourceAccessMap.entrySet()) {
+		rolesCollection = new ArrayList<GrantedAuthority>();
+		String api = entry.getKey();
+	        Map<String, Object> rolesMap = (Map<String, Object>) entry.getValue();
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java]: Object declarations moved before the loop

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java [hunk 14]
```diff
@@ -54,7 +64,58 @@ public static String extractApiKey(HttpServletRequest request) throws ApiKeyExtr
...
+	        List<String> roles = (List<String>) rolesMap.get(ROLES);
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java]: Object declarations moved before the loop

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java [hunk 15]
```diff
@@ -54,7 +64,58 @@ public static String extractApiKey(HttpServletRequest request) throws ApiKeyExtr
...
+		for(String role : roles) {
+		    rolesCollection.add(new SimpleGrantedAuthority(role));
+		}
+		authenticationToken = 
+		    new EuropeanaAuthenticatonToken(rolesCollection);
```
[the remark this answers is not in the document]
Comment (rgraf) [reply commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/OAuthUtils.java]: Constractor parameters are now: grantedAuthorities, details, and principal

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java (hunk 1 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java]: processJwtToken method removed from interface

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java (hunk 2 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/AuthorizationService.java]: authorizeWriteAccess gets request as input parameter

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 3 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: Currently, there is no "apidemo" check anymore.

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 4 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: request object is used as an input now

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 5 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: changed to protected

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 6 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: extracted method isOperationSupported. Object declarations moved before the loop.

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 7 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: Added "operation not authorized" exceptions. Return type changed to void.

File: commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/Operations.java (hunk 8 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/Operations.java]: api specific permissions removed from the file

File: commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java (hunk 9 above)
Comment (rgraf) [reply commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java]: User Roles from entity-web are used and suggested methods implemented in api commons in interface.

File: commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
+package eu.europeana.api.commons.oauth2.utils;
```
Comment (gsergiu) [inline commons-oauth/src/main/java/eu/europeana/api/commons/oauth2/utils/EuropeanaAuthenticatonToken.java]: move this class to package eu.europeana.api.commons.oauth2.model.impl

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 5 above)
Comment (gsergiu) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: the method must not be static

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 6 above)
Comment (gsergiu) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: the methodName should be changed to isAccessAuthorized
the new method  isOperationSupported must not be static

File: commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java (hunk 7 above)
Comment (gsergiu) [reply commons-web/src/main/java/eu/europeana/api/commons/service/authorization/BaseAuthorizationService.java]: refactor code 
  if(authenticationApi.equals(api)) {
		 res = isOperationSupported(operation, authorityList, userRoles);
	     }
	     
	     if(res) {
		 break;
	     }


to 
if(authenticationApi.equals(api))  & isOperationSupported(operation, authorityList, userRoles){
 return;
}

File: commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java (hunk 9 above)
Comment (gsergiu) [reply commons-web/src/main/java/eu/europeana/api/commons/web/model/vocabulary/UserRoles.java]: we must not use plural in class names, please reactor 
eu.europeana.api.commons.definitions.vocabulary.Roles to eu.europeana.api.commons.definitions.vocabulary.Role

**Pull request merged**